Allow default alert contact to be configured

aequitas opened the pull request:
- Convert alert contact ID to string (it can be 0 padded number)
- Add alert contact datasource
- Extend example configuration with default alert contact

TODO:

- [x] Pagination
- [x] Migration code for int to string conversion alert contact id
- [x] Tests

drubin commented on uptimerobot/data_source_uptimerobot_alert_contact.go:
```diff
@@ -0,0 +1,48 @@
+		return fmt.Errorf("Failed to find alert contact with name %s", friendlyName)
+	}
+
+	d.Set("friendly_name", alertContact.FriendlyName)
```
Should we not add the `id` here?

drubin commented on uptimerobot/api/alert_contact.go:
```diff
@@ -37,6 +38,44 @@ type AlertContact struct {
+func (client UptimeRobotApiClient) GetAlertContacts() (acs []AlertContact, err error) {
+	data := url.Values{}
+
+	body, err := client.MakeCall(
```
How do we handle pagination here?

aequitas replied:
This is now implemented.

aequitas commented:
I added a configuration example of what I think this will look like when fully implemented. I'll look into the pagination stuff as well.

aequitas commented:
The latest commit is a working example. I still need to add pagination support. 

I had to convert the alert contact ID from int to string as the value can be a zero padded number and the correct alert contact will not be found without those zero's. Maybe this applies to the other resources as well? Also I have not tested this with an existing configuration so migration code might need to be added as well.

aequitas commented:
So as far as I can tell there is no reason for migration, as Terraform stored the ID's as strings in the state file. So by reading them as string instead of int in the new code they are automatically read correctly.

leeif commented on uptimerobot/api/alert_contact.go:
```diff
@@ -94,12 +144,12 @@ func (client UptimeRobotApiClient) CreateAlertContact(req AlertContactCreateRequ
 	}
 
-	return client.GetAlertContact(int(alertcontact["id"].(float64)))
+	return client.GetAlertContact(alertcontact["id"].(string))
```
The alertcontact["id"] here is seems to be a float64 type which can't be converted by .(string).

aequitas replied:
According to the API documentation the ID is a string https://uptimerobot.com/api

![image](https://user-images.githubusercontent.com/365827/55827628-ef0e4080-5b0a-11e9-9d2b-e552d5d71f14.png)

I stumbled upon this because one of the ID's generated for me by the API started with a `0` which would be stripped when converting form a string to a int/float64 and thus Terraform would record an invalid ID in the state file.

leeif replied:
May be this is a uptimerobot issure. At least for me, the response of this API is different from the document.
https://gist.github.com/leeif/0bc4eb5f9ca03b9c487169d98167cc40

aequitas replied:
I think this is an issue at UR's end.

Alert contact create returns a integer:

```
$ curl -sS -X POST -H "Cache-Control: no-cache" -H "Content-Type: application/x-www-form-urlencoded" -d "api_key=XXXX&format=json&friendly_name=Test2&type=2&value=test2%40example.com"  "https://api.uptimerobot.com/v2/newAlertContact"|jq .
{
  "stat": "ok",
  "alertcontact": {
    "id": 2809628
  }
}
```

While getAlertContacts returns a string:
```
$ curl -sS -X POST -H "Cache-Control: no-cache" -H "Content-Type: application/x-www-form-urlencoded" -d "api_key=XXXXXX&format=json"  "https://api.uptimerobot.com/v2/getAlertContacts"|jq .
{
  "stat": "ok",
  "offset": 0,
  "limit": 50,
  "total": 3,
  "alert_contacts": [
    {
      "id": "0717807",
      "friendly_name": "uptimerobottest@ijohan.nl",
      "type": 2,
      "status": 2,
      "value": "uptimerobottest@ijohan.nl"
    },
    {
      "id": "2806958",
      "friendly_name": "Test",
      "type": 2,
      "status": 0,
      "value": "test@example.com"
    },
    {
      "id": "2809628",
      "friendly_name": "Test2",
      "type": 2,
      "status": 0,
      "value": "test2@example.com"
    }
  ]
}
```

One thing I do notice is that the ID of the default alert contact starts with 0 but the others are in a completely different number range.

aequitas replied:
Also querying without the zero does not work:
```
$ curl -sS -X POST -H "Cache-Control: no-cache" -H "Content-Type: application/x-www-form-urlencoded" -d "api_key=XXXXX&format=json&alert_contacts=717807"  "https://api.uptimerobot.com/v2/getAlertContacts"|jq .
{
  "stat": "fail",
  "error": {
    "type": "not_found",
    "message": "Alert Contact(id: 717807) not found."
  }
}
```

```
$ curl -sS -X POST -H "Cache-Control: no-cache" -H "Content-Type: application/x-www-form-urlencoded" -d "api_key=XXXXX&format=json&alert_contacts=0717807"  "https://api.uptimerobot.com/v2/getAlertContacts"|jq .
{
  "stat": "ok",
  "offset": 0,
  "limit": 50,
  "total": 1,
  "alert_contacts": [
    {
      "id": "0717807",
      "friendly_name": "uptimerobottest@ijohan.nl",
      "type": 2,
      "status": 2,
      "value": "uptimerobottest@ijohan.nl"
    }
  ]
}
```

aequitas replied:
So I added a conversion to string instead of casting. I don't think UR will change their API and break all clients out there (didn't attempt to contact them though). But this should cover both cases:
- in `newAlertContact` where the ID is returned as a integer
- in all other cases where it is handled as a string

Also only the response values are types (due to JSON) and the request values not (as they are HTTP parameters).

louy commented:
Nice work @aequitas!

Is this ready to be merged? Do you still want to add a test?

aequitas commented:
@louy I added a test and fixed some bugs I found when running the other acceptance tests. I also did a small testrun 'converting' a configuration created using the latest master to one with this branch and found no issues. So afaict this now works as expected and can be merged.
